dpipe/itertools.py

```diff
@@ -129,24 +129,41 @@ def __init__(self, func: Callable, iterable: Iterable, *args, **kwargs) -> None:
         self.__working_thread = Thread(
             target = self._prediction_func
         )
+        self.__exhausted = False
 
     def start(self) -> None:
         self.__working_thread.start()
 
+    def stop(self) -> None:
+        self.__working_thread.join()
+        assert not self.__working_thread.is_alive()
+        self.__exhausted = True
+
     def _prediction_func(self) -> None:
-        for value in self.__iterable:
-            self.__result_queue.put(self.__func(value, *self.__args, **self.__kwargs))
-        self.__result_queue.put(FinishToken)
+        try:
+            for value in self.__iterable:
+                self.__result_queue.put(self.__func(value, *self.__args, **self.__kwargs))
+            self.__result_queue.put(FinishToken)
+        except BaseException as e:
+            self.__result_queue.put(e)
```

Comment:
in theory the thread might _return_ the exception, so it's better to add a success flag: (value, True) or (error, False) in the queue, depending on the case.